torch/lib/THCS/THCSTensor.cu

```diff
@@ -1,22 +1,238 @@
 #include "THCSTensor.h"
 #include "THCApply.cuh"
+#include "THCTensorMathPointwise.cuh"
 #include "stdio.h"
 
-template <typename IndexType, typename Real>
-__global__ void THCSTensor_toDenseKernel(
-    TensorInfo<Real, IndexType> other,
-    TensorInfo<long, IndexType> indices,
+template <typename IndexType, typename Real, typename Op>
+__device__ void applyOp2(
+    Op op, IndexType blockSize,
+    TensorInfo<Real, IndexType> values1, IndexType idx1,
+    TensorInfo<Real, IndexType> values2, IndexType idx2) {
+  for (IndexType k = blockIdx.x * blockDim.x + threadIdx.x;
+       k < blockSize;
+       k += gridDim.x * blockDim.x) {
+    op(values1.data + idx1 * blockSize + k, values2.data + idx2 * blockSize + k);
+  }
+}
+
+template <typename IndexType, typename Real, typename Op>
+__device__ void applyOp3(
+    Op op, IndexType blockSize,
+    TensorInfo<Real, IndexType> values1, IndexType idx1,
+    TensorInfo<Real, IndexType> values2, IndexType idx2,
+    TensorInfo<Real, IndexType> values3, IndexType idx3) {
+  for (IndexType k = blockIdx.x * blockDim.x + threadIdx.x;
+       k < blockSize;
+       k += gridDim.x * blockDim.x) {
+    op(values1.data + idx1 * blockSize + k,
+       values2.data + idx2 * blockSize + k,
+       values3.data + idx3 * blockSize + k);
+  }
+}
+
+template <typename Op, typename IndexType, typename Real>
+__global__ void THCSTensor_spcKernel(
```

Comment:
can we please call it `sparseElementwiseKernel`? It's so hard to remember what `spc` means. Same for the one below.